File: backend/app/schemas/dataset.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, field_validator


class DatasetColumnRead(BaseModel):
    name: str
    label: str
    question_id: str | None = None
    kind: str


class DatasetRead(BaseModel):
    form_id: str
    mode: str
    total_rows: int
    total_columns: int
    columns: list[DatasetColumnRead]
    rows: list[dict[str, Any]]


class DataDictionaryItemRead(BaseModel):
    question_id: str
    column_name: str
    code: str | None
    label: str
    question_type: str
    question_role: str
    measurement_level: str
    data_type: str
    is_required: bool
    is_scored: bool
    is_reverse_scored: bool
    section_title: str | None
    instrument_name: str | None
    dimension_name: str | None
    project_variable_name: str | None
    options: list[dict[str, Any]]


class DataDictionaryRead(BaseModel):
    form_id: str
    items: list[DataDictionaryItemRead]


class CompletenessItemRead(BaseModel):
    question_id: str
    column_name: str
    label: str
    total_responses: int
    answered_count: int
    missing_count: int
    missing_percent: float
    required: bool
    warning_level: str


class CompletenessSummaryRead(BaseModel):
    form_id: str
    total_responses: int
    items: list[CompletenessItemRead]


class DatasetExportRequest(BaseModel):
    mode: str = Field(default="mixed")
    include_metadata: bool = True
    include_discarded: bool = False
    expand_multiple_choice: bool = False
    include_scores: bool = False


class DatasetExportRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    project_id: str
    form_id: str | None
    artifact_type: str
    file_name: str
    file_path: str
    mime_type: str | None
    file_size_bytes: int | None
    metadata_json: dict | list | None
    created_at: datetime


class DatasetExportListResponse(BaseModel):
    items: list[DatasetExportRead]
    total: int


class AnswerUpdateRequest(BaseModel):
    option_id: str | None = None
    value_text: str | None = None
    value_number: float | None = None
    value_date: datetime | None = None
    value_json: list[str] | list[dict] | dict | None = None

    @field_validator("value_text")
    @classmethod
    def normalize_optional_text(cls, value: str | None) -> str | None:
        if value is None:
            return None
        cleaned = value.strip()
        return cleaned or None


class AnswerUpdateRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    response_id: str
    question_id: str
    option_id: str | None
    value_text: str | None
    value_number: float | None
    value_date: datetime | None
    value_json: dict | list | None
    score_value: float | None
    updated_at: datetime


class ResponseStatusUpdateRequest(BaseModel):
    status: str

    @field_validator("status")
    @classmethod
    def validate_status(cls, value: str) -> str:
        cleaned = value.strip()
        if cleaned not in {"partial", "complete", "discarded"}:
            raise ValueError("status must be partial, complete, or discarded")
        return cleaned


class ResponseStatusRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    response_id: str
    status: str
    updated_at: datetime

File: frontend/src/types/dataset.ts
export interface DatasetColumn {
  name: string;
  label: string;
  question_id?: string | null;
  kind: string;
}

export interface DatasetPreview {
  form_id: string;
  mode: string;
  total_rows: number;
  total_columns: number;
  columns: DatasetColumn[];
  rows: Record<string, unknown>[];
}

export interface CompletenessItem {
  question_id: string;
  column_name: string;
  label: string;
  total_responses: number;
  answered_count: number;
  missing_count: number;
  missing_percent: number;
  required: boolean;
  warning_level: string;
}

export interface CompletenessSummary {
  form_id: string;
  total_responses: number;
  items: CompletenessItem[];
}

export interface DatasetExportArtifact {
  id: string;
  project_id: string;
  form_id?: string | null;
  artifact_type: string;
  file_name: string;
  file_path: string;
  mime_type?: string | null;
  file_size_bytes?: number | null;
  metadata_json?: unknown;
  created_at: string;
}

export interface DatasetExportList {
  items: DatasetExportArtifact[];
  total: number;
}

export interface DataDictionaryOption {
  id: string;
  label: string;
  value?: string | null;
  score?: number | null;
  sort_order: number;
}

export interface DataDictionaryItem {
  question_id: string;
  column_name: string;
  code?: string | null;
  label: string;
  question_type: string;
  question_role: string;
  measurement_level: string;
  data_type: string;
  is_required: boolean;
  is_scored: boolean;
  is_reverse_scored: boolean;
  section_title?: string | null;
  instrument_name?: string | null;
  dimension_name?: string | null;
  project_variable_name?: string | null;
  options: DataDictionaryOption[];
}

export interface DataDictionary {
  form_id: string;
  items: DataDictionaryItem[];
}

// Edición de celda en la grilla de captura manual (upsert).
export interface AnswerUpsertPayload {
  raw_value?: string | null;
  clear?: boolean;
  option_id?: string | null;
  value_text?: string | null;
  value_number?: number | null;
  value_date?: string | null;
}

export interface AnswerUpsertResult {
  id: string;
  response_id: string;
  question_id: string;
  option_id?: string | null;
  value_text?: string | null;
  value_number?: number | null;
  value_date?: string | null;
  value_json?: unknown;
  score_value?: number | null;
  updated_at: string;
}
